jobs/microk8s/utils.py

```diff
@@ -0,0 +1,213 @@
+    # This should have been caught at the very beginning.
+    assert False
+
+
+def get_gh_releases():
+    """Get all releases from GH.
+
+    Returns the parsed json object or None on failure
+    """
+    releases_url = "https://api.github.com/repos/kubernetes/kubernetes/releases"
+    r = requests.get(releases_url)
+    if r.status_code == 200:
+        releases = json.loads(r.content.decode().strip())
+        return releases
+    else:
+        None
+
+
+def get_latest_pre_release(track, patch):
+    """
+    Get the latest release for track and patch
+    :param track: Something line 1.15, 1.16
+    :param patch: alpha, beta, or rc
+    :return: None if no pre-release is found of the latest pre-release name
+    """
+    releases = get_gh_releases()
+    if not releases:
+        return None
+
+    search_version = "v{}.0-{}".format(track, patch)
+    release_names = []
+    for release in releases:
+        if release['name'].startswith(search_version):
+            release_names.append(release['name'][1:])
+
+    if len(release_names) > 0:
+        max_release = release_names[0]
+        for release_candidate in release_names:
+            if compare_releases(max_release, release_candidate) < 0:
+                max_release = release_candidate
+        return max_release
+    else:
+        return None
+
+
+if __name__ == "__main__":
+    """Tests..."""
+    versions = [
+        ('1.5.3', '1.3.4', 1),
+        ('1.5.3', '1.6.4', 2),
+        ('1.5.3', '1.5.3', 0),
+        ('1.15.3', '1.15.4', 2),
+        ('1.15.3', '1.15.2', 1),
+        ('2.5.3', '3.3.4', 2),
+        ('4.5.3', '3.3.4', 1),
+        ('1.15.0-alpha.0', '1.15.0-alpha.0', 0),
+        ('1.15.0-alpha.0', '1.15.0-alpha.1', 2),
+        ('1.15.0-alpha.2', '1.15.0-alpha.1', 1),
+        ('1.15.2-alpha.2', '1.15.0-alpha.1', 1),
+        ('1.15.0-alpha.2', '1.15.2-alpha.1', 2),
+        ('1.15.0-alpha.2', '1.15.2-alpha.2', 2),
+        ('1.15.2-alpha.2', '1.15.0-beta.1', 1),
+        ('1.15.0-alpha.2', '1.15.2-rc.1', 2),
+        ('1.15.0-alpha.2', '1.15.2', 2),
+        ('4.5.3', '3.3.0-alpha.2', 1),
+    ]
+    for pair in versions:
+        res = compare_releases(pair[0], pair[1])
+        if res > 0:
+            print("{} newer than {}".format(pair[0], pair[1]))
+            assert pair[2] == 1
+        elif res < 0:
+            print("{} older than {}".format(pair[0], pair[1]))
+            assert pair[2] == 2
+        else:
+            print("{} and {} are equal".format(pair[0], pair[1]))
+            assert pair[2] == 0
+
+    releases = get_latest_pre_release("1.15","alpha")
+    print(releases)
+
+    releases = get_latest_pre_release("1.14","rc")
+    print(releases)
+
+    releases = get_latest_pre_release("1.12","rc")
+    print(releases)
+
+    snap = Microk8sSnap("1.15", "edge", juju_controller="aws-us-east-1:default", juju_unit="ubuntu/0")
+    snap.build_and_release("1.15.0-alpha.2")
```

Comment:
Not sure if you're interested, but i took a different approach, rather than having all these versions hardcoded, i just take the tag releases from upstream and make sure they are synced in the snap store

https://github.com/charmed-kubernetes/jenkins/blob/master/jobs/build-snaps/snaps-source.py#L36-L77

Create branches for the ones we support and dont exist in launchpad and just have them built automatically. 

Also, now that I think about it, you could just use re-use some of the snaps I build from source and set your snapcraft to use stage-snaps, https://docs.snapcraft.io/snapcraft-parts-metadata (see `stage-snaps`)

Reply:
This was misleading, sorry. This part of the code was only used for local testing. Removed.